core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1118,10 +1174,40 @@ class ReplicaManager(val config: KafkaConfig,
         highWatermarkCheckpoints(dir).write(hwms)
       } catch {
         case e: IOException =>
-          fatal("Error writing to highwatermark file: ", e)
-          Exit.halt(1)
+          error("Error writing to highwatermark file: ", e)
+          handleLogDirFailure(dir)
+      }
+    }
+  }
+
+  def handleLogDirFailure(dir: String) {
+    info(s"Stopping serving replicas in dir $dir")
+
+    replicaStateChangeLock synchronized {
+      val newOfflinePartitions = allPartitions.values.filter { partition =>
+        partition.getReplica(config.brokerId) match {
+          case Some(replica) =>
+            replica.log.isDefined && replica.log.get.dir.getParentFile.getAbsolutePath == dir
+          case None => false
+        }
+      }.map(_.topicPartition)
+
+      info(s"Partitions ${newOfflinePartitions.mkString(",")} are offline due to failure on log directory $dir")
+
+      newOfflinePartitions.foreach { topicPartition =>
+        val partition = allPartitions.remove(topicPartition)
+        partition.removePartitionMetrics()
+        val topicHasPartitions = allPartitions.keys.exists(tp => topicPartition.topic == tp.topic)
+        if (!topicHasPartitions)
+          brokerTopicStats.removeMetrics(topicPartition.topic)
       }
+      replicaFetcherManager.removeFetcherForPartitions(newOfflinePartitions.toSet)
+      highWatermarkCheckpoints = highWatermarkCheckpoints.filterKeys(_ != dir)
+      stateChangeLogger.warn("Broker %d stopped fetcher for partitions %s because they are in the failed log dir %s"
```

Comment:
Previously I thought this is used for making partition state change. They are the same because previously all state change are triggered by controller request. I have changed the code to log it in the server.log.